iele-binary.md

```diff
@@ -118,41 +118,53 @@ contains enough information to determine and decode the rest of the operation.
-
-    syntax Bool ::= #isValidContract(WordStack)              [function]
-                  | #isValidContract(WordStack, Int)         [function, klabel(isValidContractAux)]
-                  | #isValidStringTable(WordStack, Int, Int) [function]
- // -------------------------------------------------------------------
-    rule #isValidContract(CODE) => #isValidContract(CODE, #sizeWordStack(CODE))
-    rule #isValidContract(.WordStack, 0) => true
-    rule #isValidContract(W1 : W2 : W3 : W4 : 99 : NBITS : WS, SIZE) => #fun(DECLSIZE => SIZE -Int 4 >=Int DECLSIZE andBool #isValidStringTable(#take(DECLSIZE -Int 2, WS), NBITS, DECLSIZE -Int 2))(W1 *Int 16777216 +Int W2 *Int 65536 +Int W3 *Int 256 +Int W4)
-    rule #isValidContract(_, _) => false [owise]
-
-    rule #isValidStringTable(105 : W1 : W2 : WS, NBITS, SIZE) => SIZE -Int 3 >=Int W1 *Int 256 +Int W2 andBool #isValidStringTable(#drop(W1 *Int 256 +Int W2, WS), NBITS, SIZE -Int 3 -Int W1 *Int 256 -Int W2)
-    rule #isValidStringTable(106 : W1 : W2 : WS, NBITS, SIZE) => SIZE -Int 3 >=Int W1 *Int 256 +Int W2 andBool #isValidContract(#take(W1 *Int 256 +Int W2, WS), W1 *Int 256 +Int W2) andBool #isValidStringTable(#drop(W1 *Int 256 +Int W2, WS), NBITS, SIZE -Int 3 -Int W1 *Int 256 -Int W2)
-      requires W1 =/=Int 0 orBool W2 =/=Int 0
-    rule #isValidStringTable(106 : 0 : 0 : WS, NBITS, SIZE) => false
-    rule #isValidStringTable(WS, NBITS, SIZE) => #isValidFunctions(WS, NBITS, SIZE) [owise]
+    syntax Contract ::= #dasmContract ( Bytes , IeleName )                                                                       [function]
+                      | #dasmContract ( start: Int , width: Int , bytecode: Bytes , name: IeleName )                             [function]
+                      | #dasmContract ( start: Int , width: Int , bytecode: Bytes , nbits: Int , functions: Map,name: IeleName ,
+                                        definitions: TopLevelDefinitions, funcnum: Int , size: Int , bytecodestring: String )    [function]
+ // ---------------------------------------------------------------------------------------------------------------------------------------
+    rule #dasmContract( BS, NAME ) => #dasmContract( 0, lengthBytes(BS), BS, NAME )
+    rule #dasmContract( _, 0, BS, NAME ) => #emptyCode
+    rule #dasmContract( I, N, BS, NAME ) => #dasmContract(I +Int 6, #asUnsigned(I, 4, BS) -Int 2, BS, BS[I +Int 5], 0 |-> init, NAME, .TopLevelDefinitions, 1, #asUnsigned(I, 4, BS) +Int 4, Bytes2String(BS[I .. N]))
+      requires N >=Int 5 andBool BS[I +Int 4] ==Int 99
+
+    rule #dasmContract( I, N, BS, NBITS, FUNCS, NAME, DEFS, M, SIZE, BYTES )
+      => #dasmContract( I +Int 3 +Int #asUnsigned(I +Int 1, 2, BS), N -Int (3 +Int #asUnsigned(I +Int 1, 2, BS)), BS, NBITS, M |-> String2IeleName(Bytes2String(BS[I +Int 3 .. #asUnsigned(I +Int 1, 2, BS)])) FUNCS, NAME, DEFS, M +Int 1, SIZE, BYTES )
+      requires N >=Int 3 andBool BS[I] ==Int 105
```

Comment:
Is it intended  that this will fail if we have two functions named the same things? (because we're using the constructor `M |-> ... FUNCS` rather than `FUNCS [ M <- ... ]`)

Reply:
I didn't write that part, but it must be. A iele contract is malformed if it has multiple functions or global definitions with the same name.

Reply:
Do we have a test for this somewhere in the test suite that tests that the transaction reachen termination with a malformed contract exit code?

Reply:
M gets incremented every time this map is updated, so actually I guess this can't fail. I'm assuming the check for malformedness happens somewhere else, in `well-formedness.md` maybe?

Reply:
Correct

Reply:
It looks like the `illFormed` test is checking for a status code of `0x09`, I can't find a reference to that number right away though.

Reply:
That's the correct exit code; if there is a test for multiple functions with the same name, we can probably resolve this.

Reply:
I don't believe a test for this exists.